### ISSUE TITLE
Possible use of MONAI?

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/ivado-medical-imaging#211]
[This project](https://github.com/Project-MONAI/MONAI) might have relevant codebase that could be used within ivadomed, such as loss functions, transforms, architectures, etc.

**Comment by charleygros** [comment]
Something that I think could be relevant is the way they handle the transforms.

For instance: they define transforms for images, and other for labels (eg [here](https://github.com/Project-MONAI/MONAI/blob/f87a258326c8f74f35653e4b305a9642ec4e4356/examples/segmentation_3d_ignite/unet_evaluation_array.py#L50)).
Then, in the loader, the im_transforms and seg_transforms are called separately: [here](https://github.com/Project-MONAI/MONAI/blob/e3e79fce0529d777a42dab2f23b1c6d4e7fcca9e/monai/data/nifti_reader.py#L117).
--> which could be mimicked [here in our code](https://github.com/neuropoly/ivado-medical-imaging/blob/master/ivadomed/loader/loader.py#L356).

Pros
- we could remove all the `if self.labeled` and `if "roi" in sample` in `transforms.py`
- the transforms that are applied to the labels are explicit

**Comment by uzaymacar** [comment]
## Summary of discussion

MONAI was mentioned again recently in an ivadomed meeting. To summarize the discussion and findings:

* MONAI provides building blocks to write your own code whereas ivadomed is more structured with a config file.
* MONAI data loader is built up on PyTorch dataloader whereas ivadomed adopts the BIDS data structure on top.
* Recently we integrated `torchio` in ivadomed, which provides more medical-specific augmentations than MONAI.
* There aren't many differences between ivadomed and MONAI w.r.t. loss functions. In fact, ivadomed includes an implementation for `AdapWingLoss` whereas MONAI doesn't.
* We can refactor our `models.py` and use MONAI's modeling, that is their strength. Specifically, they provide the following models: UNet, DynUNet, DenseNet, GAN, AHNet, VNet, SENet(and SEResNet, SEResNeXt), SegResNet, EfficientNet, and attention-based transformers.

## Preliminary exploration

Focusing the preliminary exploration on U-Net models, we see that MONAI's [`BasicUnet`](https://github.com/Project-MONAI/MONAI/blob/a676e3876903e5799eb3d24245872ea4f3a5b152/monai/networks/nets/basic_unet.py#L189) corresponds to and covers our [`Unet`](https://github.com/ivadomed/ivadomed/blob/b5d67218d9a172b44fe4d9b2bb3fdcfd2a10a725/ivadomed/models.py#L472) and [`Modified3DUNet`](https://github.com/ivadomed/ivadomed/blob/b5d67218d9a172b44fe4d9b2bb3fdcfd2a10a725/ivadomed/models.py#L745) implementations. 

Implementation-wise, we'll just have to match the **already existing keys** in the config file to the following common parameters found in MONAI's model classes:

```
spatial_dims: number of spatial dimensions. Defaults to 3 for spatial 3D inputs.
in_channels: number of input channels. Defaults to 1.
out_channels: number of output channels. Defaults to 2.
features: six integers as numbers of features. Defaults to ``(32, 32, 64, 128, 256, 32)``,
act: activation type and arguments. Defaults to LeakyReLU.
norm: feature normalization type and arguments. Defaults to instance norm.
bias: whether to have a bias term in convolution blocks. Defaults to True.
dropout: dropout ratio. Defaults to no dropout.
upsample: upsampling mode, available options are ``"deconv"``, ``"pixelshuffle"``, ``"nontrainable"``.
```

## Discussion

However, it is important to note that MONAI's implementations do not have the option to add attention layers, whereas ivadomed does. MONAI has a separate [`AttentionUnet`](https://github.com/Project-MONAI/MONAI/blob/a676e3876903e5799eb3d24245872ea4f3a5b152/monai/networks/nets/attentionunet.py#L164) for this. Moreover, FiLM layers and HeMIS are not implemented in MONAI whereas these are integrated within the [`Unet`](https://github.com/ivadomed/ivadomed/blob/b5d67218d9a172b44fe4d9b2bb3fdcfd2a10a725/ivadomed/models.py#L472) and [`Modified3DUNet`](https://github.com/ivadomed/ivadomed/blob/b5d67218d9a172b44fe4d9b2bb3fdcfd2a10a725/ivadomed/models.py#L745) in ivadomed. 

In this scenario, it might be better if we integrate **segmentation** models from MONAI that **don't exist in ivadomed** instead of completely refactoring `models.py`. For example, attention-based transformers (e.g. UNETR) could be nice to integrate as they are popular and newcomers will be potentially interested in them. 

What do people think? Any other suggestions?
